base/reduce.jl

```diff
@@ -439,6 +439,9 @@ mapreduce(f, op, a::Number) = mapreduce_first(f, op, a)
 
 _mapreduce(f, op, ::IndexCartesian, A::AbstractArrayOrBroadcasted) = mapfoldl(f, op, A)
 
+@inline _mapreduce(f, op, ::IndexCartesian, A::AbstractVector) =
+    _mapreduce(f, op, IndexLinear(), A)
```

Comment:
We can handle 1D Broadcasted too

```suggestion
function _mapreduce(f, op, ::IndexCartesian, A::AbstractArrayOrBroadcasted)
    if IteratorSize(A) isa HasShape
        if ndims(A) == 1
            return _mapreduce(f, op, IndexLinear(), A)
        end
    end
    return mapfoldl(f, op, A)
end
```

Reply:
I guess we don't need this. As all one-dimensional `Broadcasted` is linear-indexed if their `ndims` has been defined.
Only `AbstractVector` might been cartesian-indexed.
https://github.com/JuliaLang/julia/blob/a32a0660dfe28442b3b3e823021462ebbeb90729/base/broadcast.jl#L242-L248

Reply:
I think it'd be more productive to mark the discussion resolved once we arrive at a conclusion.

Notice that your reasoning requires that `mapfoldl(f, op, A::Broadcasted)` dispatches to a `getindex`-based implementation. I just checked that this is not the case with

```julia
julia> bc = Broadcast.instantiate(Broadcast.broadcasted(+, [1,2,3], 4));

julia> @descend mapfoldl(identity, +, bc)
```

Reply:
Sorry for my recklessness. But I'm affraid I didn't catch you.
For your example `IndexStyle(bc) === IndexLinear()`, thus it won't hit `_mapreduce(f, op, ::IndexCartesian, A::AbstractArrayOrBroadcasted)`. Am I right?

Reply:
Thanks for the clarification. So you were saying that it's the caller who specifies the correct index style for `bc` with `ndim(bc) == 1`. Yes, indeed that seems to be the case.

But then this raises another question: why do we need to enforce the index style of `_mapreduce(f, op, ::IndexCartesian, A::AbstractVector)`? Arguably, it is possible that `typeof(A)` provides an efficient `iterate` than `getindex` which is why it's not declared to be `IndexLinear` in the first place. I'm not sure if SparseArrays.jl does this but I'd guess you can build a state machine for `iterate` on a sparse vector that is faster than `getindex`. It applies to other non-trivial vector types too. If we call `IndexLinear` here, we are disabling the optimization done by the type implementer.

So, my suggestion now is to simply remove `_mapreduce(f, op, ::IndexCartesian, A::AbstractVector)`

```suggestion
_mapreduce(f, op, ::IndexCartesian, A::AbstractArrayOrBroadcasted) = mapfoldl(f, op, A)
```

Reply:
I do the above change mainly because our `fold` messes up at speed some time.
```julia
julia> a = randn(4096);
julia> @btime foldl(Base.add_sum,$a);
  3.688 μs (0 allocations: 0 bytes)
julia> @btime reduce(Base.add_sum,$a);
  296.935 ns (0 allocations: 0 bytes)
```
> Arguably, it is possible that typeof(A) provides an efficient iterate than getindex which is why it's not declared to be IndexLinear in the first place.

Our user defined vectors are default to be `IndexCartesian`.
And it's also a little confused to me that `IndexCartesian` means `iterate` is more efficient than `getindex`.
Anyway, I'll remove it.